File: fast_slow_pointer/llist_cycle.cpp
#include <iostream>
#include <vector>

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
   public:
    bool hasCycle(ListNode *head) {
        ListNode *sptr = head, *fptr = head;

        while (fptr && fptr->next) {
            sptr = sptr->next;
            fptr = fptr->next->next;

            if (fptr == sptr) {
                return true;
            }
        }

        return false;
    }
};

// case null
// case 1

int main() {}

File: fast_slow_pointer/llist_cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        # create a set, each time a node is visted add to list
        # if node is already in set, there is circle
        record = set()

        ptr = head
        while ptr != None:
            if ptr in record:
                return True
            record.add(ptr)
            ptr = ptr.next
        return False
